zepben-extensions/rmqstream/src/lib.rs

```diff
@@ -0,0 +1,177 @@
+    }
+
+    debug!("Reading in parameters from C types...");
+    let host = CStr::from_ptr(_host).to_string_lossy().to_string();
+    let port = _port as u16;
+    let user = CStr::from_ptr(_user).to_string_lossy().to_string();
+    let pass = CStr::from_ptr(_pass).to_string_lossy().to_string();
+    let stream = CStr::from_ptr(_stream).to_string_lossy().to_string();
+    let heartbeat = _heartbeat as u32;
+
+    debug!("C params read. Connecting to RabbitMQ stream ({user}@{host}:{port}, stream {stream})...");
+
+    let producer = RUNTIME.block_on(async {
+        // Retry connection up to 3 times
+        let mut retries = 0;
+        let max_retries = 3;
+        let mut last_error = None;
+
+        while retries < max_retries {
+            match Environment::builder()
+                .host(&host)
+                .port(port)
+                .heartbeat(heartbeat)
+                .username(&user)
+                .password(&pass)
+                .build()
+                .await
+            {
+                Ok(environment) => {
+                    debug!("Connected. Making producer...");
+                    match environment
+                        .producer()
+                        .batch_size(100000)
+                        .batch_delay(Duration::from_millis(250))
+                        .build(&stream)
+                        .await
+                    {
+                        Ok(producer) => return producer,
+                        Err(e) => {
+                            last_error = Some(e.to_string());
+                            warn!("Failed to create producer (attempt {} of {}): {}", retries + 1, max_retries, e);
+                        }
+                    }
+                },
+                Err(e) => {
+                    last_error = Some(e.to_string());
+                    warn!("Connection failed (attempt {} of {}): {}", retries + 1, max_retries, e);
+                }
+            }
+
+            retries += 1;
+            if retries < max_retries {
+                sleep(Duration::from_secs(1)).await; // Wait before retrying
+            }
+        }
+
+        panic!(
+            "Could not connect to RabbitMQ after {} attempts. Last error: {:?}",
+            max_retries,
+            last_error
+        );
+    });
+
+    PRODUCER = Some(producer);
+    TOTAL_MESSAGES = 0;
+    TOTAL_BYTES = 0;
+    START_TIME = Some(Instant::now());
+    info!(
+        "Connected to RabbitMQ {}@{}:{}, for stream '{}'.",
+        &user, &host, port, &stream
+    );
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn disconnect_from_stream() {
+    if let Some(producer) = PRODUCER.take() {
+        RUNTIME.block_on(async {
+            // There seems to be a bug in the rust stream client lib: outgoing messages are not
+            // confirmed to have been sent before attempting to close the producer. We wait
+            // 2 seconds here to be safe.
+            sleep(Duration::from_secs(2)).await;
```

Comment:
If two seconds it not enough... what happens? Will be get an error or data loss?

Reply:
data loss iirc. I'll try the send_with_confirm thing that this library has for the final message.